internal/cmd/agent.go

```diff
@@ -36,27 +30,40 @@ func newAgentCmd() *cobra.Command {
 		RunE: func(cmd *cobra.Command, _ []string) error {
 			infraDir, err := initInfraHomeDir()
 			if err != nil {
-				panic(err)
+				return fmt.Errorf("infra home directory: %w", err)
 			}
-
 			logging.UseFileLogger(filepath.Join(infraDir, "agent.log"))
 
-			var wg sync.WaitGroup
-
-			// start background tasks
-			ctx, cancel := context.WithCancel(context.Background())
-			defer cancel()
-
-			ctx = context.WithValue(ctx, countCtxKey, &counter{count: 0, threshold: 3})
-
-			repeat.InGroup(&wg, ctx, cancel, 1*time.Minute, syncKubeConfig)
+			group, ctx := errgroup.WithContext(context.Background())
+			group.Go(func() error {
+				backOff := &backoff.ExponentialBackOff{
+					InitialInterval:     time.Minute,
+					MaxInterval:         5 * time.Minute,
+					RandomizationFactor: 0.05,
+					Multiplier:          1.5,
+					// MaxElapsedTime is the duration the infra agent should
+					// continue to run without a successful sync. After this time
+					// the agent will exit.
+					MaxElapsedTime: 15 * time.Minute,
```

Comment:
Instead of a fixed 3 failures, this new implementation does the following:
1. starts exponential backoff from 1 minute up to 5 minutes
2. after 15 minutes of no successful operations it will exit

Does 15 minutes seem like a reasonable delay? I'm happy to change these settings.

It uses a 5% randomization to prevent thundering herds when recovering from an outage, and a 1.5 multiplier to add time each interval.

Reply:
I didn't bother setting an exponential back off for this since 1 minute intervals are infrequent enough that consecutive failures won't cause excessive noise. The problem with exponential back off is that more than a few failures produces a too long wait between attempts. At that point, it's likely better to exit

Reply:
I'm hoping we'll be able to use long polling here pretty soon, so that we can significantly decrease the wait  time between operations.  Until then, I could reduce the MaxInterval to 2 minutes, and change toe MaxElapsedTime to ~5 minutes. That should be pretty close to the old behaviour.

Reply:
Ok, I've adjusted these values to be closer to the previous behaviour.